modeling/algorithm/shared-global-algorithm-core.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package algorithm
+
+import (
+	"errors"
+	"math"
+
+	"github.com/googleinterns/k8s-topology-simulator/modeling/types"
+)
+
+// SharedGlobalAlgorithmCore takes multiple zones as input and output
+// EntpointSliceGroups composition based on their nodes and endpoints:
+// 1. EndpointSlices will be considered global by default
+// 2. Once the number of endpoints in a zone reaches a specified threshold,
+//	  zone-specific EndpointSlices will begin to be produced.
+// 3. Kube Proxy will consume zone specific EndpointSlices. And decide whether
+//    to consume the global EndpointSlice based on users input.
+type SharedGlobalAlgorithmCore struct {
+	// Weight of global EndpointSliceGroup
+	globalWeight float64
+	// Threshold of global EndpointSliceGroup that if the total number of endpoints
+	// <= threshold, all endpoints go to global EndpointSliceGroup
+	globalThreshold int
+}
+
+// CreateSliceGroups takes a region of zones as input and output
+// EndpointSliceGroups
+func (alg SharedGlobalAlgorithmCore) CreateSliceGroups(region types.RegionInfo, excludeContributor bool) (map[string]types.EndpointSliceGroup, error) {
+	if region.ZoneDetails == nil {
+		return nil, errors.New("can't create EndpointSlices with 0 number of zone")
+	}
+	// The deviation for the traffic and capacity above
+	deviation := make(map[string]float64)
+
+	for _, zone := range region.ZoneDetails {
+		// Calculate the deviation based on the capacity(endpoints) and
+		// traffic(nodes) ratio
+		deviation[zone.Name] = float64(zone.Endpoints) - float64(region.TotalEndpoints)*zone.NodesRatio
+	}
+
+	// Output EndpointSlices
+	sliceGroups := make(map[string]types.EndpointSliceGroup)
+	// The global sliceGroup -- might be split into many small global slices
```

Comment:
Had a similar question above, what does `--` mean here?

Reply:
sorry for the legacy comment here, updated to a more reasonable one